client/src/main/java/io/pravega/client/stream/ReaderGroupConfig.java

```diff
@@ -183,25 +250,41 @@ public ReaderGroupConfig build() {
                endingStreamCuts = Collections.emptyMap();
            }
 
+           //fetch StreamCuts with the specified default scope.
+           final Map<String, StreamCut> startCutsWithDefaultScope = getCutsWithDefaultScope(startingStreamCuts);
+           final Map<String, StreamCut> endCutsWithDefaultScope = getCutsWithDefaultScope(endingStreamCuts);
+
            //validate start and end StreamCuts
-           validateStreamCut(startingStreamCuts);
-           validateStreamCut(endingStreamCuts);
+           validateStreamCut(startCutsWithDefaultScope);
+           validateStreamCut(endCutsWithDefaultScope);
 
            //basic check to verify if endStreamCut > startStreamCut.
-           validateStartAndEndStreamCuts(startingStreamCuts, endingStreamCuts);
+           validateStartAndEndStreamCuts(startCutsWithDefaultScope, endCutsWithDefaultScope);
 
-           return new ReaderGroupConfig(groupRefreshTimeMillis, automaticCheckpointIntervalMillis,
-                   startingStreamCuts, endingStreamCuts);
+           return new ReaderGroupConfig(groupRefreshTimeMillis, automaticCheckpointIntervalMillis, defaultScope,
+                   startCutsWithDefaultScope, endCutsWithDefaultScope);
        }
 
-       private void validateStartAndEndStreamCuts(Map<Stream, StreamCut> startStreamCuts,
-                                                  Map<Stream, StreamCut> endStreamCuts) {
+        private Map<String, StreamCut> getCutsWithDefaultScope(final Map<String, StreamCut> streamCutMap) {
+            return streamCutMap.entrySet().stream().collect(Collectors.toMap(e -> {
+                Stream s = Stream.of(e.getKey());
+                if (s.getScope().equals(USE_DEFAULT_SCOPE)) {
```

Comment:
This feels hacky. We should not have to do a string comparison to use a default scope. Why can't we assume that the streams in the cut already have the right scope? I don't see a need for a stream in the cut to have `_defaultScope` as its scope so that we can disambiguate it lazily here.

Reply:
My current reasoning to use this approach is as follows, do let me know your thoughts on this.

i) The default scope to be used by the ReaderGroupConfiguration is set by using `ReaderGroupConfigBuilder#defaultScope`

ii) The following two user patterns can be used by the user.
```

 ReaderGroupConfig.builder()
                  .defaultScope("scope")
                  .addStream("streamA", startStreamCut1) //add stream scope/streamA
                  .build();
OR	

     ReaderGroupConfig.builder()
                  .addStream("streamA", startStreamCut1) // add stream scope/streamA
                  .defaultScope("scope")
                 .build();
```

Both the patterns are valid and this implies at the time `addStream` is invoked we do not know the defaultScope to be used.
	
iii) The `ReaderGroupConfig.builder().addStream(...)` internally stores the mapping of `Stream.getScopedName()` to the associatied start and end streamCut.
(`Map<String, StreamCut>startingStreamCuts` and `Map<String, StreamCut> endingStreamCuts`)

The above points imply that during ReaderConfigBuilder.build() we need to ensure that the mapping of  `startingStreamCuts` and `endingStreamCuts` use the scope
set by `defaultScope`.
(This is currently performed by `getCutsWithDefaultScope(..)` method and we have a Precondition to ensure defaultScope() is set incase the .addStream() is invoked without passing the scope. )

Reply:
We faced a similar situation in the Flink connector: the user provides a string (which we call a stream spec) which may or may not be qualified, and we must defer the task of qualifying the string for a while.   @shrids is it true that `build()` has enough info to qualify the string into a typed `Stream`?   Is this notion of a placeholder scope name encapsulated in the builder (as a detail of its internal data structure)?

Reply:
> Both the patterns are valid and this implies at the time addStream is invoked we do not know the defaultScope to be used.

True, but we do know at build time, so we could fill in the default scope if needed at build time. Also, isn't it possible that `startFromCheckpoint` is invoked before `defaultScope`, in which case we would have a problem as `getCutsWithDefaultScope` expects the default scope to be set in the case `s.getScope().equals(USE_DEFAULT_SCOPE)` holds?

Reply:
`ReaderGroupConfigBuilder#startFromCheckpoint` updates the `Map<String, StreamCut>startingStreamCuts` and since a checkpoint will not have `USE_DEFAULT_SCOPE`  as the scope in its Stream to StreamCut mapping `getCutsWithDefaultScope()` will not try to replace it with the default scope; `ReaderGroupConfigBuilder#defaultScope()` will be a no-op for `startFromCheckPoint(..)` case.